cli/azd/pkg/ux/canvas.go

```diff
@@ -0,0 +1,94 @@
+package ux
+
+import (
+	"io"
+	"sync"
+)
+
+// Canvas is a base component for UX components that require a canvas for rendering.
+type canvas struct {
+	visuals    []Visual
+	printer    Printer
+	writer     io.Writer
+	renderMap  map[Visual]*VisualContext
+	updateLock sync.Mutex
+}
+
+type Canvas interface {
+	Run() error
+	Update() error
+	WithWriter(writer io.Writer) Canvas
+}
+
+// NewCanvas creates a new Canvas instance.
+func NewCanvas(visuals ...Visual) Canvas {
+	canvas := &canvas{
+		visuals:   visuals,
+		renderMap: make(map[Visual]*VisualContext),
+	}
+
+	return canvas
+}
+
+// WithWriter sets the writer for the canvas.
+func (c *canvas) WithWriter(writer io.Writer) Canvas {
+	c.writer = writer
+	return c
+}
+
+// Run starts the canvas.
+func (c *canvas) Run() error {
+	c.printer = NewPrinter(c.writer)
+	return c.Update()
+}
+
+// Update updates the canvas to force a re-render.
+func (c *canvas) Update() error {
+	c.updateLock.Lock()
+	defer c.updateLock.Unlock()
+
+	if c.printer == nil {
```

Comment:
Just wondering, what does the `updateLock` protect here?

At a glance, it's not immediately clear what the actual critical resource we're trying to protect here, and if this lock is necessary.

FWIW, it's really hard to reason about critical resources here since we have separated components based on abstract responsibilities rather than control flow. I do prefer when the underlying implementation is immediately clear, such as in bubbletea's [renderer](https://github.com/charmbracelet/bubbletea/blob/main/standard_renderer.go).

Reply:
The `updateLock` prevents multiple re-renders of a component at the same time.